File: design-patterns/go/visitor/hosting_cost/hosting_calculator_visitor_impl.go
// hosting_calculator_visitor_impl.go

package main

type HostingCalculatorVisitorImpl struct{}

func NewHostingCalculatorVisitorImpl() (hcvi *HostingCalculatorVisitorImpl) {
	hcvi = &HostingCalculatorVisitorImpl{}
	return
}
func (hcvi *HostingCalculatorVisitorImpl) ComputeVisit(computeService *ComputeService) float64 {
	return computeService.GetPrice() * float64(computeService.GetQuantity())
}

func (hcvi *HostingCalculatorVisitorImpl) ContainerVisit(containerService *ContainerService) float64 {
	return containerService.GetPrice() * float64(containerService.GetQuantity())
}

func (hcvi *HostingCalculatorVisitorImpl) DatabaseVisit(databaseService *DatabaseService) float64 {
	serviceCost := databaseService.GetPrice() * float64(databaseService.GetQuantity())

	backupCost := 0.00
	if databaseService.IsBackupEnabled() {
		backupCost = databaseService.GetBackPrice() * float64(databaseService.GetQuantity())
	}
	return serviceCost + backupCost
}

func (hcvi *HostingCalculatorVisitorImpl) FileVisit(fileStorageService *FileStorageService) float64 {
	return fileStorageService.GetPricePerGB() * float64(fileStorageService.GetQuantity())
}

func (hcvi *HostingCalculatorVisitorImpl) ServerlessVisit(serverlessService *ServerlessService) float64 {
	return serverlessService.GetHourlyPrice() * serverlessService.GetHourlyPrice()
}

File: design-patterns/php/visitor/hostingcost/HostingCalculatorVisitorImpl.php
<?php
// HostingCalculatorVisitorImpml.php

namespace BigBoxCode\DesignPattern\Visitor\HostingCost;

class HostingCalculatorVisitorImpl implements HostingCalculatorVisitor {
    public function visitComputeService(ComputeService $computeService): float {
        return $computeService->getPrice() * $computeService->getQuantity();
    }

    public function visitDatabaseService(DatabaseService $databaseService): float {
        $serviceCost = $databaseService->getPrice() * $databaseService->getQuantity();
        $backupCost = 0;

        if ($databaseService->isBackupEnabled()) {
            $backupCost = $databaseService->getBackupPrice() * $databaseService->getQuantity();
        }

        return $serviceCost + $backupCost;
    }

    public function visitFileStorageService(FileStorageService $fileStorageService): float {
        return $fileStorageService->getPricePerGB() * $fileStorageService->getQuantity();
    }

    public function visitServerlessService(ServerlessService $serverlessService): float {
        return $serverlessService->getHourlyPrice() * $serverlessService->getTotalHours();
    }

    public function visitContainerService(ContainerService $containerService): float {
        return $containerService->getPrice() * $containerService->getQuantity();
    }
}